src/java.base/share/classes/java/lang/reflect/InvocationHandler.java

```diff
@@ -96,7 +101,10 @@ public interface InvocationHandler {
      *
      * @see     UndeclaredThrowableException
      */
-    public Object invoke(Object proxy, Method method, Object[] args)
+    @CFComment("nullness: A null return is allowed only if the Method points to a method whose"
+        + " return value may be null. We don't know whether it does, so we conservatively omit"
+        + " @Nullable on the return type")
+    public Object invoke(Object proxy, Method method, @Nullable Object @Nullable [] args)
```

Comment:
```suggestion
    public @Nullable Object invoke(Object proxy, Method method, Object [] args)
```

This would seem like the most conservative annotations: The parameter array is non-null, b/c for a method with parameters, it needs to be non-null. A subclass can "widen" the parameter types to allow empty parameter lists. The reverse is not possible, so this seems more conservative.
The array component type similarly should be non-null, although that would be a problem with invariant arrays.

Opposite for the return type: we don't know whether the method returns null or not, so we should make the return type nullable. If a particular subclass knows more about the return type, it can refine it to be non-null. Again, the opposite override is not possible, so this seems preferable.

I think that's what I would do if we actually type-check all callers. I think your point is that there are callers that are not running the Nullness Checker, so we want to force implementers of this interface to always handle null values.
That makes invocations of `invoke` more dangerous: we allow `null` even though that might be illegal.
However, it forces implementers of this interface to check for `null`.

If that is your argument, I see the benefit to that style for the parameter type. What is the argument for the return type?

Reply:
> I think your point is that there are callers that are not running the Nullness Checker, so we want to force implementers of this interface to always handle null values.

Oh, I hadn't been thinking of it in those terms. That's helpful, thanks.

(I also admit that I hadn't been thinking of the fact that a subclass could widen the parameters.)

A further point is that, not only are we not checking callers to ensure that they don't pass null, but we _know_ that they sometimes _do_ pass null. That's true for both the array and its elements:

- The JDK itself passes a null _array_ for a no-arg method: As far as I can tell, that's what it always does in practice. _And_ I read the Javadoc as suggesting that it always does (though it's not 100% clear).
- The JDK can pass us null _elements_ from our "real callers," who include anyone who calls the interface method that is being handled by the `InvocationHandler`. The fact is that they can treat the parameter types as nullable, so I see consistency in doing the same in our signature.

The second argument applies similarly to the return type: Our "real callers" are callers of arbitrary interface methods, which might have non-nullable return types (even primitive return types!). The fact is that they can treat the return type as non-nullable, so I see consistency in doing the same in our signature.

All that might make my argument a _little_ more principled, but I'm still mostly arguing on pragmatics rather than on principle :) So I can still go with your signature if you like.

You asked about inside Google: I think I'm expecting to use the current JSpecify signature:

[`public @Nullable Object invoke(Object proxy, Method method, @Nullable Object @Nullable[] args)`](https://github.com/jspecify/jdk/blob/9c5377b5fcd24b535c8ca3ef7c60e53b6cb29bf2/src/java.base/share/classes/java/lang/reflect/InvocationHandler.java#L97C5-L97C99)